.github/monitor-typescript-errors/utils/xml.js

```diff
@@ -0,0 +1,66 @@
+const { XMLParser } = require( 'fast-xml-parser' );
+
+exports.parseXml = ( filePath ) => {
+	const parser = new XMLParser( {
+		ignoreAttributes: false,
+		attributeNamePrefix: '',
+		attributesGroupName: '',
+	} );
+	const parsedFile = parser.parse( filePath );
+
+	return getDataFromParsedXml( parsedFile );
+};
+
+const getErrorInfo = ( error ) => {
+	const line = error.line;
+	const column = error.column;
+	const message = error.message;
+
+	return `${ line }:${ column } - ${ message }`;
+};
+
+const getDataFromParsedXml = ( parsedXml ) => {
+	const data = parsedXml.checkstyle.file;
+
+	return data.reduce(
+		( acc, { name, error } ) => {
+			const pathFile = name;
+			const hasMultipleErrors = Array.isArray( error );
+
+			return {
+				files: {
+					[ pathFile ]: hasMultipleErrors
+						? error.map( getErrorInfo )
+						: [ getErrorInfo( error ) ],
+					...acc.files,
+				},
+				totalErrors:
+					acc.totalErrors + ( hasMultipleErrors ? error.length : 1 ),
+				totalFilesWithErrors: acc.totalFilesWithErrors + 1,
+			};
+		},
+		{
+			totalErrors: 0,
+			totalFilesWithErrors: 0,
+		}
+	);
+};
+
+exports.getFilesWithNewErrors = (
+	newCheckStyleFileParsed,
+	currentCheckStyleFileParsed
+) => {
+	const newFilesReport = newCheckStyleFileParsed.files;
+	const currentFilesReport = currentCheckStyleFileParsed.files;
+
+	return Object.keys( newFilesReport ).reduce(
```

Comment:
Would it be a good idea to sort these alphabetically so we don't mix folders?
Also instead of displaying a list of just files, could it be more useful to display a list of errors, eg file + line?

While we're at it, hashing the file+lines combos and comparing that instead could be a way of getting rid of the "solving one, introduce another" problem you mentioned. I'm out of ideas for dealing with file renaming though 🤔

Reply:
> Would it be a good idea to sort these alphabetically, so we don't mix folders?

Agreed. I added the sort.

> Also instead of displaying a list of just files, could it be more useful to display a list of errors, eg file + line:
folder/filename:<error_line_number>


I agree, but it is not so easy to implement, and I don't see that it makes sense to make this effort. As the code is structured now, if I add the error_line_number, it will be added all the errors of that file, not only the errors that the developer introduced in the PR. Furthermore, if I implement an algorithm that checks only the difference between the old `checkstyle.xml` and the new one with ` files&lines combos` when the developers add a space line, the algorithm will fail because all the lines are changing, so for this latter, new errors are introduced even if it isn't true.

Reply:
I see. Since we're dealing with tech debt here, can this job not be merge-blocking for now (although you can force merge stuff)?

Maybe produce a warning about typescript errors in the code, or to keep it failing, analyze only changed files (as the git hook does for phpcs).

We're suffering a bit with flaky tests, another red mark on the pipeline would desensitize us even more and cause blind force merges 🤔

Reply:
It is expected behavior that the CI fails on this branch:

```
Now, the action is failing because the branch to compare is the trunk (that doesn't have a checkstyle.xml).
```